### PR TITLE
[DPE-3881] Use ruff as a formatter and linter

### PR DESCRIPTION
**Pull request by shayancanonical**
## Issue
We are using black as a formatter. We would like to use ruff

## Solution
Use ruff as a linter and formatter. Run format on the repo and push up changes as a result of the format

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -36,22 +36,13 @@ opentelemetry-exporter-otlp-proto-http = "1.21.0"
...
 [tool.poetry.group.lint.dependencies]
-black = "^24.0.0"
-isort = "^5.12.0"
-flake8 = "^7.0.0"
-flake8-docstrings = "^1.7.0"
-flake8-copyright = "^0.2.4"
-flake8-builtins = "^2.1.0"
-pyproject-flake8 = "^7.0.0"
-pep8-naming = "^0.13.3"
-codespell = "^2.2.5"
```
[the remark this answers is not in the document]
Comment (paulomach) [reply pyproject.toml:54]: It's. If we want to go all in rust tools, [typos](https://pypi.org/project/typos/) is very good

**Review by taurus-forever — COMMENTED** [review]
+ check CI tests: https://github.com/canonical/mysql-k8s-operator/actions/runs/9939511290/job/27454234922#step:4:19

File: pyproject.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml:54]: added back codespell in ce263127a446b0358b73a5c7c5f6fe99a1148a5a

File: pyproject.toml
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml:1]: replaced flake8 config with equivalent rust config in ce263127a446b0358b73a5c7c5f6fe99a1148a5a

File: pyproject.toml [hunk 2]
```diff
@@ -93,9 +84,10 @@ markers = ["unstable", "juju3", "only_with_juju_secrets", "only_without_juju_sec
...
 # Formatting tools configuration
-[tool.black]
+[tool.ruff]
 line-length = 99
-target-version = ["py38"]
+exclude = ["lib"]
+target-version = "py38"
 
 [tool.isort]
 profile = "black"
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: removed in ce263127a446b0358b73a5c7c5f6fe99a1148a5a

File: pyproject.toml [hunk 3]
```diff
@@ -92,30 +84,44 @@ log_cli_level = "INFO"
...
-# Linting tools configuration
-[tool.flake8]
-max-line-length = 99
-max-doc-length = 99
-max-complexity = 10
-exclude = [".git", "__pycache__", ".tox", "build", "dist", "*.egg_info", "venv"]
-select = ["E", "W", "F", "C", "N", "R", "D", "H"]
-# Ignore W503, E501 because using black creates errors with this
+[tool.ruff.lint]
+explicit-preview-rules = true
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml:96]: `explicit-preview-rules` does not prefix match in the select list. therefore, only `CPY001` will be selected from the preview rules - the others are prefixes except `I001` which is stable

File: pyproject.toml [hunk 4]
```diff
@@ -92,30 +84,44 @@ log_cli_level = "INFO"
...
+select = ["A", "E", "W", "F", "C", "N", "D", "I001", "CPY001"]
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: sounds like a good idea - enabled in 0cfb80c9db9d1c60757ac389e965425c6b7b941f

File: pyproject.toml [hunk 5]
```diff
@@ -92,30 +84,44 @@ log_cli_level = "INFO"
...
+extend-ignore = [
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: merged with ignore in 99e2ef34f167302cc6ce180bce4482dbeb653040

File: pyproject.toml [hunk 6]
```diff
@@ -92,30 +84,44 @@ log_cli_level = "INFO"
...
+# Ignore E501 because using  creates errors with this
 # Ignore D107 Missing docstring in __init__
-ignore = ["W503", "E501", "D107"]
-# D100, D101, D102, D103: Ignore missing docstrings in tests
-per-file-ignores = ["tests/*:D100,D101,D102,D103,D104"]
-docstring-convention = "google"
+ignore = ["E501", "D107"]
+
+[tool.ruff.lint.per-file-ignores]
+"tests/*" = ["D100", "D101", "D102", "D103", "D104"]
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply pyproject.toml]: addressed in 99e2ef34f167302cc6ce180bce4482dbeb653040

**Review by taurus-forever — APPROVED** [review]

**Pull request merged**